src/debugpy/server/tracing/__init__.py

```diff
@@ -0,0 +1,533 @@
+                return frame
+        return StackFrame(thread, frame_object)
+
+    @classmethod
+    def get(self, id: int) -> "StackFrame":
+        return self._all.get(id, None)
+
+    def scopes(self) -> List[Scope]:
+        if self._scopes is None:
+            self._scopes = [
+                Scope(self.frame_object, "local"),
+                Scope(self.frame_object, "global"),
+            ]
+        return self._scopes
+
+    @classmethod
+    def invalidate(self, thread_id: int):
+        frames = [frame for frame in self._all.values() if frame.thread.id == thread_id]
+        VariableContainer.invalidate(*frames)
+
+
+@dataclass
+class Step:
+    step: Literal["in", "out", "over"]
+    origin: FrameType = None
+    origin_line: int = None
+
+
+class Condition:
+    """
+    Expression that must be true for the breakpoint to be triggered.
+    """
+
+    expression: str
+    """Python expression that must evaluate to True for the breakpoint to be triggered."""
+
+    _code: CodeType
+
+    def __init__(self, breakpoint: "Breakpoint", expression: str):
+        self.expression = expression
+        self._code = compile(
+            expression, f"breakpoint-{breakpoint.id}-condition", "eval"
+        )
+
+    def test(self, frame: StackFrame) -> bool:
+        """
+        Returns True if the breakpoint should be triggered in the specified frame.
+        """
+        try:
+            return bool(
+                eval(
+                    self._code,
+                    frame.frame_object.f_globals,
+                    frame.frame_object.f_locals,
+                )
+            )
+        except:
+            log.exception(
+                f"Exception while evaluating breakpoint condition: {self.expression}"
+            )
+            return False
+
+
+class HitCondition:
+    """
+    Hit count expression that must be True for the breakpoint to be triggered.
+
+    Must have the format `[<operator>]<count>`, where <count> is a positive integer literal,
+    and <operator> is one of `==` `>` `>=` `<` `<=` `%`, defaulting to `==` if unspecified.
+
+    Examples:
+        5: break on the 5th hit
+        ==5: ditto
+        >5: break on every hit after the 5th
+        >=5: break on the 5th hit and thereafter
+        %5: break on every 5th hit
+    """
+
+    _OPERATORS = {
+        "==": lambda expected_count, count: count == expected_count,
+        ">": lambda expected_count, count: count > expected_count,
+        ">=": lambda expected_count, count: count >= expected_count,
+        "<": lambda expected_count, count: count < expected_count,
+        "<=": lambda expected_count, count: count <= expected_count,
+        "%": lambda expected_count, count: count % expected_count == 0,
+    }
+
+    hit_condition: str
+    _count: int
+    _operator: Callable[[int, int], bool]
+
+    def __init__(self, hit_condition: str):
+        self.hit_condition = hit_condition
+        m = re.match(r"([<>=]+)?(\d+)", hit_condition)
```

Comment:
```suggestion
        m = re.match(r"([%<>=]+)?(\d+)", hit_condition)
```

Reply:
Since it's checking the operators against a predefined list on the next step anyway, I'll simplify it down to any sequence of non-digits followed by digits.